sycl/doc/design/GroupSort.md

```diff
@@ -0,0 +1,220 @@
+
+DPC++ Headers contain the following:
+- required definitions of `joint_sort`, `sort_over_group` functions, `radix_order` enum class,
+  `default_sorter`, `radix_sorter` classes with corresponding `operator()`
+  as well as other classes and methods.
+
+- Checks if radix sort is applicable for provided data types.
+
+- Traits to distinguish interfaces with `Compare` and `Sorter` parameters.
+
+- Fallback solution for user's types, user's comparators and/or user's sorters.
+
+### Level Zero
+
+To implement `memory_required` methods for sorters we need to calculate
+how much temporary memory is needed.
+However, we don't have an information how much memory is needed by backend compiler.
+That's why we need a Level Zero function that calls a function from the backend and
+provide actual value to the SYCL code.
+
+Required interfaces:
+```cpp
+    // Returns whether default work-group or sub-group sort is present in builtins
+    virtual bool DefaultGroupSortSupported(GroupSortMemoryScope::MemoryScope_t scope,
+                                           GroupSortKeyType::KeyType_t keyType,
+                                           bool isKeyValue,
+                                           bool isJointSort) const;
+
+    // Returns required amount of memory for default joint work-group or sub-group sort
+    // devicelib builtin function in bytes per workgroup (or sub-group), >= 0
+    // or -1 if the algorithm for the specified parameters is not implemented
+    //
+    // totalItems -- number of elements to sort
+    // rangeSize -- work-group or sub-group size respectively
+    //
+    // For key-only sort pass valueTypeSizeInBytes = 0
+    virtual long DefaultGroupJointSortMemoryRequired(GroupSortMemoryScope::MemoryScope_t scope,
+                                                     long totalItems,
+                                                     long rangeSize,
+                                                     long keyTypeSizeInBytes,
+                                                     long valueTypeSizeInBytes) const;
+
+    // Returns required amount of memory for default private memory work-group or sub-group sort
+    // devicelib builtin function in bytes per workgroup (or sub-group), >= 0
+    // or -1 if the algorithm for the specified parameters is not implemented
+    //
+    // itemsPerWorkItem -- number of elements in private array to sort
+    // rangeSize -- work-group or sub-group size respectively
+    //
+    // For key-only sort pass valueTypeSizeInBytes = 0
+    virtual long DefaultGroupPrivateSortMemoryRequired(GroupSortMemoryScope::MemoryScope_t scope,
+                                                       long itemsPerWorkItem,
+                                                       long rangeSize,
+                                                       long keyTypeSizeInBytes,
+                                                       long valueTypeSizeInBytes) const;
+```
+
+### Fallback SPIR-V library
```

Comment:
It is expected that device compiler implements those functions as part of some "extension" so that SYCL RT can query if there is native support for that functionality and link fallback libraries if there is not. See [extension spec, which is more of a design doc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/C-CXX-StandardLibrary.rst#device-compiler).

What should be the name of this "library"/"extension"? Should there be several of them so we can only link-in those libraries which are actually used (in case they would be huge)?